packages/pds/src/app-migrations/delete-derefed-blobs.ts

```diff
@@ -0,0 +1,122 @@
+import AppContext from '../context'
+import { cborToLex } from '@atproto/repo'
+import { Record as ProfileRecord } from '../lexicon/types/app/bsky/actor/profile'
+import { AtUri } from '@atproto/uri'
+import { dedupe } from '../labeler/util'
+import { chunkArray } from '@atproto/common'
+import { CID } from 'multiformats/cid'
+import { ImageUriBuilder } from '../image/uri'
+import Database from '../db'
+
+export async function migration(ctx: AppContext) {
+  const deletedDbBlobs = await ctx.db.transaction(async (dbTxn) => {
+    await derefedRecords(dbTxn)
+    await derefedProfiles(dbTxn)
+    const { ref } = dbTxn.db.dynamic
+
+    const deleted = await dbTxn.db
+      .deleteFrom('blob')
+      .whereNotExists(
+        dbTxn.db
+          .selectFrom('repo_blob')
+          .whereRef('repo_blob.did', '=', ref('blob.creator'))
+          .whereRef('repo_blob.cid', '=', ref('blob.cid'))
+          .selectAll(),
+      )
+      .returningAll()
+      .execute()
+    return deleted
+  })
+
+  const deletedDbBlobCids = dedupe(deletedDbBlobs.map((row) => row.cid))
+
+  const stillExistingChunks = await Promise.all(
+    chunkArray(deletedDbBlobCids, 100).map(async (chunk) => {
+      const res = await ctx.db.db
+        .selectFrom('blob')
+        .where('cid', 'in', chunk)
+        .select('cid')
+        .execute()
+      return res.map((row) => row.cid)
+    }),
+  )
+  const stillExisting = stillExistingChunks.flat()
+
+  const toDeleteBlobs = deletedDbBlobCids.filter(
+    (cid) => !stillExisting.includes(cid),
+  )
+
+  const chunks = chunkArray(toDeleteBlobs, 100)
+  for (const chunk of chunks) {
+    await Promise.all([
+      ...chunk.map((cid) => ctx.blobstore.quarantine(CID.parse(cid))),
+      ...chunk.map((cid) => {
+        const paths = ImageUriBuilder.commonSignedUris.map((id) => {
+          const uri = ctx.imgUriBuilder.getCommonSignedUri(id, cid)
+          return uri.replace(ctx.imgUriBuilder.endpoint, '')
+        })
+        return ctx.imgInvalidator?.invalidate(cid, paths)
+      }),
+    ])
+  }
+}
+
+async function derefedRecords(dbTxn: Database) {
+  dbTxn.assertTransaction()
+
+  const { ref } = dbTxn.db.dynamic
+  await dbTxn.db
+    .deleteFrom('repo_blob')
+    .whereNotExists(
+      dbTxn.db
+        .selectFrom('record')
+        .whereRef('record.uri', '=', ref('repo_blob.recordUri'))
+        .selectAll(),
+    )
+    .returningAll()
+    .execute()
+}
+
+async function derefedProfiles(dbTxn: Database) {
+  dbTxn.assertTransaction()
```

Comment:
We might find this is heavy for a single transaction.  If that's an issue, even just doing it in batches by did might do the trick.